src/Admin.js

```diff
@@ -0,0 +1,88 @@
+import React, { Children, Component } from 'react';
+import PropTypes from 'prop-types';
+import { Admin as OriginalAdmin } from 'admin-on-rest';
+import AuthProvider from './AuthProvider';
+import { AUTH_GET_PERMISSIONS } from './constants';
+import defaultResolvePermissions, { resolvePermission as defaultResolvePermission } from './resolvePermissions';
+
+export const filterResourcesFromWithPermission = async ({
+    authClient,
+    withPermission,
+    resolvePermission = defaultResolvePermission,
+}) => {
+    const permissions = await authClient(AUTH_GET_PERMISSIONS);
+    const match = await resolvePermission({ permissions })({
+        exact: withPermission.props.exact,
+        permissions: withPermission.props.value,
+        resolve: withPermission.props.resolve,
+    });
+
+    return match.matched ? withPermission.props.children : false;
+};
+
+export const filterResourcesFromSwitchPermissions = async ({
+    authClient,
+    switchPermissions,
+    resolvePermissions = defaultResolvePermissions,
+}) => {
+    const mappings =
+        Children.map(switchPermissions.props.children, ({ props: { value, resolve, children, exact } }) => ({
+            permissions: value,
+            resolve,
+            view: Children.toArray(children),
+            exact,
+        })) || [];
+
+    const permissions = await authClient(AUTH_GET_PERMISSIONS);
+    const match = await resolvePermissions({ mappings, permissions });
+
+    return match ? match.view : false;
+};
+
+export const filterResourcesFromPermissions = async ({ authClient, resources: allResources }) => {
+    const resources = await Promise.all(
+        Children.toArray(allResources).map(async child => {
+            if (child.type.displayName === 'WithPermission') {
```

Comment:
Agreed. As discussed, we should refactor the `Resource` component accordingly so that this PR and #22 became useless as using the permissions components will **just work ®**